### PR TITLE
Adds closets to ant decomposition blacklist, converts ant locate() spam into typecache

### PR DESCRIPTION
**Pull request by BurgerLUA**
## About The Pull Request

Food no longer is subject to ant decomposition in closets, lockers, crates, chests, ect.
All the bad locate() calls are now replaced with typecache.

## Why It's Good For The Game

I think this was an oversight considering that racks and conveyors are immune, but open fridges, lockers, and crates aren't. The **DEEP lore** says that space ants are shit at climbing so this would make sense.

## Changelog
:cl: BurgerBB
fix: Fixes closets/lockers/crates/ect being subject to ant decompisition.
/:cl:

### CODE REVIEW

**Comment by BurgerLUA** [comment]
-4 for all the salt PRs on TGMC smh, if this is not technically a bugfix, close this PR so I can laugh at the idiots who thought this was a bug

File: code/datums/components/food/decomposition.dm [hunk 1]
```diff
@@ -73,6 +73,9 @@
 	if(locate(/obj/machinery/conveyor) in get_turf(food)) // Makes sure no decals spawn on disposals conveyors
 		remove_timer()
 		return
+	if(locate(/obj/structure/closet) in get_turf(food)) //Is it on a closet/fridge/locker/crate?
+		remove_timer()
+		return
```
Comment (Mothblocks) [inline code/datums/components/food/decomposition.dm:78]: Remove all these stupid checks and do what I asked them to do from the beginning.

Loop over contents of the turf *once*, and check if they're in a typecache.

We should not be performing *four* `locate in turf` checks, and I don't even know why we are doing 3.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/datums/components/food/decomposition.dm (hunk 1 above)
Comment (BurgerLUA) [reply code/datums/components/food/decomposition.dm:78]: I knew someone was going to suggest doing this. I'll do it then.

Comment (BurgerLUA) [reply code/datums/components/food/decomposition.dm:68]: Acomplished. Let me know if this is the incorrect way to do this.

File: code/datums/components/food/decomposition.dm [hunk 2]
```diff
@@ -61,18 +61,17 @@
...
-		remove_timer()
-		return
-	if(locate(/obj/machinery/conveyor) in get_turf(food)) // Makes sure no decals spawn on disposals conveyors
+	var/turf/open/T = food.loc
+
+	if(!istype(T)) //Are we actually in an open turf?
 		remove_timer()
 		return
+
+	for(var/k in T.contents) //Check contents of turf
```
Comment (Mothblocks) [inline code/datums/components/food/decomposition.dm:70]: https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/datums/components/food/decomposition.dm [hunk 3]
```diff
@@ -61,18 +61,17 @@
...
-		remove_timer() // If not, remove any active timers and return
-		return
-	if(locate(/obj/structure/table) in get_turf(food)) // Is this currently over a table?
-		remove_timer()
-		return
-	if(locate(/obj/structure/rack) in get_turf(food)) // Is it on a rack?
-		remove_timer()
-		return
-	if(locate(/obj/machinery/conveyor) in get_turf(food)) // Makes sure no decals spawn on disposals conveyors
+	var/turf/open/T = food.loc
```
Comment (Mothblocks) [inline code/datums/components/food/decomposition.dm:64]: https://github.com/tgstation/tgstation/blob/master/.github/CONTRIBUTING.md#use-descriptive-and-obvious-names

File: code/datums/components/food/decomposition.dm [hunk 4]
```diff
@@ -61,18 +61,17 @@
...
+
+	if(!istype(T)) //Are we actually in an open turf?
 		remove_timer()
 		return
+
+	for(var/k in T.contents) //Check contents of turf
+		if(GLOB.typecache_elevated_structures[k]) //Are we actually on an elevated structure?
```
Comment (Mothblocks) [inline code/datums/components/food/decomposition.dm:71]: This isn't going to work. You need to actually check the `.type`.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/_globalvars/lists/typecache.dm [hunk 5]
```diff
@@ -10,3 +10,5 @@ GLOBAL_LIST_INIT(typecache_living, typecacheof(/mob/living))
 GLOBAL_LIST_INIT(typecache_stack, typecacheof(/obj/item/stack))
 
 GLOBAL_LIST_INIT(typecache_machine_or_structure, typecacheof(list(/obj/machinery, /obj/structure)))
+
+GLOBAL_LIST_INIT(typecache_elevated_structures, typecacheof(list(/obj/structure/table,/obj/structure/rack,/obj/machinery/conveyor,/obj/structure/closet))) //This is used for ants.
```
Comment (Mothblocks) [inline code/_globalvars/lists/typecache.dm:14]: 1. Auto-doc this
2. Your doc-comment should explain what it *is*, not what something else happens to use it for.

File: code/datums/components/food/decomposition.dm (hunk 4 above)
Comment (BurgerLUA) [reply code/datums/components/food/decomposition.dm:71]: Oh this is silly of me, yeah I keep doing this on my own coderbase and I never learn.

File: code/_globalvars/lists/typecache.dm (hunk 5 above)
Comment (BurgerLUA) [reply code/_globalvars/lists/typecache.dm:14]: What is auto-doc? Other than a transformer character. Is this on github documents?

Comment (Mothblocks) [reply code/_globalvars/lists/typecache.dm:14]:
```
/// bla bla bla
var/here
```

File: code/_globalvars/lists/typecache.dm [hunk 6]
```diff
@@ -11,4 +11,5 @@ GLOBAL_LIST_INIT(typecache_stack, typecacheof(/obj/item/stack))
 
 GLOBAL_LIST_INIT(typecache_machine_or_structure, typecacheof(list(/obj/machinery, /obj/structure)))
 
-GLOBAL_LIST_INIT(typecache_elevated_structures, typecacheof(list(/obj/structure/table,/obj/structure/rack,/obj/machinery/conveyor,/obj/structure/closet))) //This is used for ants.
+//A typecache listing structures that are considered to have surfaces that you can place items on that are higher than the floor. This, of course, should be restricted to /atom/movables. This is primarily used for food decomposition code.
```
Comment (Mothblocks) [inline code/_globalvars/lists/typecache.dm:14]:
```suggestion
/// A typecache listing structures that are considered to have surfaces that you can place items on that are higher than the floor. This, of course, should be restricted to /atom/movables. This is primarily used for food decomposition code.
```

Doc comment == ///

File: code/datums/components/food/decomposition.dm [hunk 7]
```diff
@@ -61,14 +61,15 @@
...
-	if(!istype(T)) //Are we actually in an open turf?
+	if(!istype(open_turf)) //Are we actually in an open turf?
 		remove_timer()
 		return
 
-	for(var/k in T.contents) //Check contents of turf
-		if(GLOB.typecache_elevated_structures[k]) //Are we actually on an elevated structure?
+	for(var/content in open_turf.contents) //Check contents of turf
+		var/atom/movable/movable_content = content //Convert to movable.
+		if(GLOB.typecache_elevated_structures[movable_content.type]) //Are we actually on an elevated structure?
```
Comment (Mothblocks) [inline code/datums/components/food/decomposition.dm:72]:
```suggestion
	for(var/content as anything in open_turf.contents)
		if(GLOB.typecache_elevated_structures[content.type])
```

**Review by Mothblocks — CHANGES_REQUESTED** [review]
Last changes, probably.

**Review by Mothblocks — APPROVED** [review]

File: code/datums/components/food/decomposition.dm [hunk 8]
```diff
@@ -61,18 +61,17 @@
...
-		remove_timer()
-		return
-	if(locate(/obj/machinery/conveyor) in get_turf(food)) // Makes sure no decals spawn on disposals conveyors
+	var/turf/open/open_turf = food.loc
+
+	if(!istype(open_turf)) //Are we actually in an open turf?
 		remove_timer()
 		return
+
+	for(var/atom/content as anything in open_turf.contents)
```
Comment (Rohesie) [inline code/datums/components/food/decomposition.dm:70]:
```suggestion
	for(var/atom/movable/content as anything in open_turf.contents)
```
Turf contents can only be movable.

**Comment by Rohesie** [comment]
I'd much prefer the property to sit on the object itself instead of a type cache, but eh.

**Comment by Wallemations** [comment]
goddamn this is way cleaner than what I did, thanks mate.

**Pull request merged**